src/libpmemfile-posix/inode.c

```diff
@@ -626,6 +626,119 @@ inode_free(PMEMfilepool *pfp, TOID(struct pmemfile_inode) tinode)
 }
 
 /*
+ * vinode_rdlock2 -- take READ locks on specified inodes in always
+ * the same order
+ */
+void
+vinode_rdlock2(struct pmemfile_vinode *v1, struct pmemfile_vinode *v2)
+{
+	if (v1 == v2)
+		os_rwlock_rdlock(&v2->rwlock);
+	else if ((uintptr_t)v1 < (uintptr_t)v2) {
+		os_rwlock_rdlock(&v1->rwlock);
+		os_rwlock_rdlock(&v2->rwlock);
+	} else {
+		os_rwlock_rdlock(&v2->rwlock);
+		os_rwlock_rdlock(&v1->rwlock);
+	}
+}
+
+/*
+ * vinode_wrlock2 -- take WRITE locks on specified inodes in always
+ * the same order
+ */
+void
+vinode_wrlock2(struct pmemfile_vinode *v1, struct pmemfile_vinode *v2)
+{
+	if (v1 == v2)
+		os_rwlock_wrlock(&v2->rwlock);
+	else if ((uintptr_t)v1 < (uintptr_t)v2) {
+		os_rwlock_wrlock(&v1->rwlock);
+		os_rwlock_wrlock(&v2->rwlock);
+	} else {
+		os_rwlock_wrlock(&v2->rwlock);
+		os_rwlock_wrlock(&v1->rwlock);
+	}
+}
+
+/*
+ * vinode_unlock2 -- drop locks on specified inodes
+ */
+void
+vinode_unlock2(struct pmemfile_vinode *v1, struct pmemfile_vinode *v2)
+{
+	if (v1 == v2) {
+		os_rwlock_unlock(&v1->rwlock);
+	} else {
+		os_rwlock_unlock(&v1->rwlock);
+		os_rwlock_unlock(&v2->rwlock);
+	}
+}
+
+/*
+ * vinode_cmp -- compares 2 inodes
+ */
+static int
+vinode_cmp(const void *v1, const void *v2)
+{
+	return (int)((intptr_t)*(void **)v1 - (intptr_t)*(void **)v2);
+}
+
+/*
+ * vinode_in_array -- returns true when vinode is already in specified array
+ */
+static bool
+vinode_in_array(const struct pmemfile_vinode *vinode,
+		struct pmemfile_vinode * const *arr,
+		size_t size)
+{
+	for (size_t i = 0; i < size; ++i)
+		if (arr[i] == vinode)
+			return true;
+	return false;
+}
+
+/*
+ * vinode_wrlock4 -- take up to 4 WRITE locks on specified inodes in always
+ * the same order
+ */
+void
+vinode_wrlock4(struct pmemfile_vinode *v[], size_t *N,
```

Comment:
It might look weird to you, but it removes a variable. Of which, there are already too many for most of us to follow.
The compiler can check it for you.
```
int func(int x[static 5])
{
        return x[2] + x[3];
}

int v[4];
func(vv);

test.c:67:2: warning: array argument is too small; contains 4 elements, callee requires at least 5 [-Warray-bounds]
        func(vv);
```
Anyways, couldn't this be a function with variadic arguments, with a null terminated argument list? When supplied with the right GNU attributes, the compiler can also check for the existence of the null terminator.